contracts/base/FunctionHandlerManager.sol

```diff
@@ -0,0 +1,85 @@
+pragma solidity ^0.8.18;
+import {IERC165} from "@openzeppelin/contracts/utils/introspection/IERC165.sol";
+import {ISafeProtocolFunctionHandler} from "../interfaces/Integrations.sol";
+import {ISafe} from "../interfaces/Accounts.sol";
+import {RegistryManager} from "./RegistryManager.sol";
+
+/**
+ * @title FunctionHandlerManager
+ * @notice This contract manages the function handlers for the Safe Account. The contract stores the
+ *        information about Safe account, bytes4 function selector and the function handler contract address.
+ * @dev This contract inherits BaseManager so that `onlyPermittedIntegration` modifier can be used.
+ */
+abstract contract FunctionHandlerManager is RegistryManager {
+    // Storage
+    /** @dev Mapping that stores information about Safe account, function selector, and address of the account.
+     */
+    mapping(address => mapping(bytes4 => address)) public functionHandlers;
+
+    // Events
+    event FunctionHandlerChanged(address indexed safe, bytes4 indexed selector, address indexed functionHandler);
+
+    // Errors
+    error AddressDoesNotImplementFunctionHandlerInterface(address functionHandler);
+    error FunctionHandlerNotSet(address safe, bytes4 functionSelector);
+
+    /**
+     * @notice Returns the function handler for a Safe account and function selector.
+     * @param safe Address of the Safe account
+     * @param selector bytes4 function selector
+     * @return functionHandler Address of the contract to be set as a function handler
+     */
+    function getFunctionHandler(address safe, bytes4 selector) external view returns (address functionHandler) {
+        functionHandler = functionHandlers[safe][selector];
+    }
+
+    /**
+     * @notice Sets the function handler for a Safe account and function selector. The msg.sender must be the account.
+     * @param selector bytes4 function selector
+     * @param functionHandler Address of the contract to be set as a function handler
+     */
+    function setFunctionHandler(bytes4 selector, address functionHandler) external onlyPermittedIntegration(functionHandler) {
+        // No need to check if functionHandler implements expected interfaceId as check will be done when adding to registry.
+        functionHandlers[msg.sender][selector] = functionHandler;
+        emit FunctionHandlerChanged(msg.sender, selector, functionHandler);
+    }
+
+    /**
+     * @notice This fallback handler function checks if a safe (msg.sender) has a function handler enabled.
+     *         If enabled, calls handle function and returns the result back.
+     *         Currently, the handle(...) function is non-payable and has same signature for both ISafeProtocolFunctionHandler
+     *         and ISafeProtocolStaticFunctionHandler. So, ISafeProtocolFunctionHandler.handle is used even for static calls.
+     */
+    // solhint-disable-next-line no-complex-fallback
+    fallback(bytes calldata) external payable returns (bytes memory) {
```

Comment:
I think fallback function should not be `payable` because it will lead to native token being locked up in Manager. It was added here to get access to `msg.value` but I think we should reconsider how `value` parameter is useful in `handle(...)`

Reply:
with the current implementation of Safe Fallback Handlers value is always 0. This is more a future proof feature when we allow this for example in Safe v2. 

Sorry for not communicating this (was discussed with the CoW team before :see_no_evil: )

Reply:
so, we follow the implementation similar as [here](https://github.com/rndlabs/safe-contracts/blob/e53ffea42d8547c984c463bf5b7e47f156619a1f/contracts/handler/extensible/FallbackHandler.sol#L30-L40)?

i.e. non-payable fallback handler?